docs(cn): content/warnings/invalid-hook-call-warning.md into Chinese (#169)
reactjs/zh-hans.reactjs.org: content/warnings/invalid-hook-call-warning.md

---
title: 警告：非法 Hook 调用
layout: single
permalink: warnings/invalid-hook-call-warning.html
---

你能到这个页面很可能是因为你看到了以下错误信息：

> Hooks can only be called inside the body of a function component.

触发这个警告有三种常见的原因：

1. 你的 React 和 React DOM 可能**版本不匹配**。

> 2. 你可能**打破了 [Hooks 的使用规则](/docs/hooks-rules.html)**。

Reviewer: ```suggestion
2. 你可能**打破了 [Hook 的规范](/docs/hooks-rules.html)**。
```

> 2. 你可能**打破了 [Hooks 的使用规则](/docs/hooks-rules.html)**。

Reviewer: ```suggestion
2. 你可能**打破了 [Hooks 的使用规范](/docs/hooks-rules.html)**。
```

3. 你可能在同一个应用中拥有**多个 React 副本**。

让我们依次来看这些情况。

## React 和 React DOM 版本不匹配 {#mismatching-versions-of-react-and-react-dom}

> 你可能使用的是尚未支持 Hooks 的 `react-dom`（< 16.8.0）或 `react-native`（< 0.59）版本。你可以在应用文件夹中执行 `npm ls react-dom` 或 `npm ls react-native` 来检当前使用的版本。如果你发现不止一个这些依赖，可能也会产生问题（在后面的小节会进行说明）。

Reviewer: ```suggestion
你可能使用的是尚未支持 Hooks 的 `react-dom`（< 16.8.0）或 `react-native`（< 0.59）版本。你可以在应用文件夹中执行 `npm ls react-dom` 或 `npm ls react-native` 来检当前使用的版本。如果你发现不止一个相关的依赖，可能也会产生问题（在后面的小节会进行说明）。
```

> ## 打破了 Hooks 的使用规则 {#breaking-the-rules-of-hooks}

Reviewer: ```suggestion
## 打破了 Hook 的使用规范 {#breaking-the-rules-of-hooks}
```

> 你只能在**当 React 渲染函数组件时**调用 Hooks：

Reviewer: ```suggestion
你只能在**当 React 渲染函数组件时**调用 Hook：
```

* ✅ 在函数组件的顶层调用它们。
* ✅ 在[自定义 Hook](/docs/hooks-custom.html) 的顶层调用它们。

> **查看 [Hooks 的使用规则](/docs/hooks-rules.html)以了解更多内容。**

Reviewer: ```suggestion
**查看 [Hook 的使用规范](/docs/hooks-rules.html)以了解更多内容。**
```

```js{2-3,8-9}
function Counter() {
  // ✅ Good：函数组件的顶层
  const [count, setCount] = useState(0);
  // ...
}

function useWindowWidth() {
  // ✅ Good：自定义 Hook 的顶层
  const [width, setWidth] = useState(window.innerWidth);
  // ...
}
```

> 为避免困惑，在以下情况中调用 Hooks 是**不**受支持的：

Reviewer: ```suggestion
为避免困惑，在以下情况中调用 Hook 是**不**被支持的：
```

> * 🔴 不要在 class 组件中调用 Hooks。

Reviewer: ```suggestion
* 🔴 不要在 class 组件中调用 Hook。
```

class Bad3 extends React.Component {
  render() {
    // 🔴 Bad：在 class 组件内部
    useEffect(() => {})
    // ...
  }
}
```

> 你可以使用 [`eslint-plugin-react-hooks` 插件](https://www.npmjs.com/package/eslint-plugin-react-hooks)来捕捉这些错误。

Reviewer: ```suggestion
你可以使用 [`eslint-plugin-react-hooks` 插件](https://www.npmjs.com/package/eslint-plugin-react-hooks)来捕获这些错误。
```

> 注意
>

> > [自定义 Hooks](/docs/hooks-custom.html) *可能会*调用其他 Hooks（这恰好是他们的用途）。这是可行的，因为自定义 Hooks 也应该只在函数组件渲染时被调用。

Reviewer: ```suggestion
> [自定义 Hook](/docs/hooks-custom.html) *可能会*调用其他 Hook（这恰好是他们的用途）。这是可行的，因为自定义 Hook 也应该只在函数组件渲染时被调用。
```

## 重复的 React {#duplicate-react}

> 为了使 Hooks 工作，你应用代码中的 `react` 依赖和 `react-dom` 包内部的 `react` 依赖必须解析为同一个模块。

Reviewer: ```suggestion
为了使 Hook 正常工作，你应用代码中的 `react` 依赖以及 `react-dom` 的 package 内部使用的 `react` 依赖，必须解析为同一个模块。
```

> 如果这些 `react` 依赖解析为两个不同的导出对象，你就会看到本警告。这可能发生在你**意外地引入了两个 `react` 包的副本**。

Reviewer: ```suggestion
如果这些 `react` 依赖解析为两个不同的导出对象，你就会看到本警告。这可能发生在你**意外地引入了两个 `react` 的 package 副本**。
```

> 如果你用的是 Node 作为包管理工具，你可以执行以下代码来检查你的项目文件夹：

Reviewer: ```suggestion
如果你用的是 Node 作为 package 管理工具，你可以执行以下代码来检查你的项目文件夹：
```
